pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -394,6 +394,10 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.I
 		PortRange:  fmt.Sprintf("%d-%d", minPort, maxPort),
 		Target:     gce.targetPoolURL(name, region),
 	}
+	if loadBalancerIP != nil {
+		req.IPAddress = loadBalancerIP.String()
```

Comment:
externalIPs are "unmanaged".  People typically use them, for example, to
stick VM IPs in and expose a Service on a low NodePort.  I don't LIKE it as
an API, but there were enough voices complaining when we tried to take it
away...

The fundamental difference between externalIPs[] and
type=LoadBalancer+loadBalancerIP is that the latter is managed by
kubernetes and the former are not.  Kubernetes will instantiate a
loadbalancer and continually ensure that it is alive and valid for the
managed case, and do nothing at all with the unmanaged case.

Tim


On Mon, Aug 31, 2015 at 10:17 PM, Angus Lees <notifications@github.com>
wrote:

> In pkg/cloudprovider/providers/gce/gce.go
> <https://github.com/kubernetes/kubernetes/pull/13005#discussion_r38386270>
> :
>
> > @@ -394,6 +394,10 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.I
> >  		PortRange:  fmt.Sprintf("%d-%d", minPort, maxPort),
> >  		Target:     gce.targetPoolURL(name, region),
> >  	}
> > +	if loadBalancerIP != nil {
> > +		req.IPAddress = loadBalancerIP.String()
>
> This. I think my question is: Why isn't this entire change just the
> addition of:
>
> if externalIP != nil {
>         req.IPAddress = externalIP.String()
> }
>
> ... and nothing else?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13005/files#r38386270>.
>